xmake/rules/c++/modules/modules_support/gcc.lua

```diff
@@ -149,77 +149,91 @@ function generate_dependencies(target, sourcebatch, opt)
+-- generate target user header units
+function generate_user_headerunits(target, batchcmds, headerunits, opt)
+    local compinst = target:compiler("cxx")
+    local mapper_file = _get_module_mapper()
 
-                batchcmds:show_progress(opt.progress, "${color.build.object}generating.cxx.headerunit.bmi %s", headerunit.name)
-                batchcmds:vrunv(compinst:program(), table.join(compinst:compflags({target = target}), args))
+    -- get cachedirs
+    local cachedir = common.modules_cachedir(target)
 
-                batchcmds:add_depfiles(headerunit.path)
-                batchcmds:set_depmtime(os.mtime(bmifile))
-                batchcmds:set_depcache(target:dependfile(bmifile))
-            end
+    -- build headerunits
+    local projectdir = os.projectdir()
+    for _, headerunit in ipairs(headerunits) do
+        local file = path.relative(headerunit.path, projectdir)
+        local objectfile = target:objectfile(file)
+
+        local outdir
+        if headerunit.type == ":quote" then
+            outdir = path.join(cachedir, path.directory(path.relative(headerunit.path, projectdir)))
         else
-            local bmifile = path.join(stlcachedir, headerunit.name .. get_bmi_extension())
-            if _add_module_to_mapper(mapper_file, headerunit.path, path.absolute(bmifile, projectdir)) then
-                if not os.isfile(bmifile) then
-                    local args = { "-c", "-x", "c++-system-header", headerunit.name }
+            outdir = path.join(cachedir, path.directory(headerunit.path))
+        end
+        if not os.isdir(outdir) then
+            os.mkdir(outdir)
+        end
 
-                    batchcmds:show_progress(opt.progress, "${color.build.object}generating.cxx.headerunit.bmi %s", headerunit.name)
-                    batchcmds:vrunv(compinst:program(), table.join(compinst:compflags({target = target}), args))
+        local bmifilename = path.basename(objectfile) .. get_bmi_extension()
+        local bmifile = (outdir and path.join(outdir, bmifilename) or bmifilename)
+        if not os.isdir(path.directory(objectfile)) then
+            os.mkdir(path.directory(objectfile))
+        end
 
-                    batchcmds:set_depmtime(os.mtime(bmifile))
-                    batchcmds:set_depcache(target:dependfile(bmifile))
-                end
-            end
+        local args = { "-c" }
+        local headerunit_path
+        if headerunit.type == ":quote" then
+            table.join2(args, { "-I", path.directory(path.relative(headerunit.path, projectdir)), "-x", "c++-user-header", headerunit.name })
+            headerunit_path = path.join(".", path.relative(headerunit.path, projectdir))
+        elseif headerunit.type == ":angle" then
+            table.join2(args, { "-x", "c++-system-header", headerunit.name })
+            -- if path is relative then its a subtarget path
+            headerunit_path = path.is_absolute(headerunit.path) and headerunit.path or path.join(".", headerunit.path)
```

Comment:
> path.join(".", headerunit.path)

why? I think it is exactly equivalent to `headerunit_path  = headerunit_path `

Reply:
> 

because gcc prepend paths with . when it's relative, and without it it doesn't work on my tests, i think GCC do a strict string comparison and not a resolved path comparison

Reply:
ok, got it.